server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java

```diff
@@ -0,0 +1,336 @@
+                final int numberOfShards = randomIntBetween(1, maxShardsToCreate);
+                try {
+                    applyClusterState("add-more-unassigned-shards-for-" + indexName,
+                        clusterState -> addUnassignedShards(clusterState, indexName, numberOfShards));
+                } catch (Exception e) {
+                    throw new AssertionError(e);
+                } finally {
+                    remainingShards -= numberOfShards;
+                }
+            }
+        });
+        addSnapshotRestoreIndicesThread.start();
+
+        assertBusy(() -> {
+            assertThat(snapshotsInfoService.numberOfKnownSnapshotShardSizes(), equalTo(succeed.size()));
+            assertThat(snapshotsInfoService.numberOfUnknownSnapshotShardSizes(), equalTo(0));
+        });
+
+        addSnapshotRestoreIndicesThread.join();
+    }
+
+    public void testNoLongerMaster() throws Exception {
+        final InternalSnapshotsInfoService snapshotsInfoService =
+            new InternalSnapshotsInfoService(Settings.EMPTY, clusterService, () -> repositoriesService, () -> rerouteService);
+
+        when(repositoriesService.repository("_repo"))
+            .thenReturn(new FilterRepository(mock(Repository.class)) {
+                @Override
+                public IndexShardSnapshotStatus getShardSnapshotStatus(SnapshotId snapshotId, IndexId indexId, ShardId shardId) {
+                    return IndexShardSnapshotStatus.newDone(0L, 0L, 0, 0, 0L, randomNonNegativeLong(), null);
+                }
+            });
+
+        for (int i = 0; i < randomIntBetween(1, 10); i++) {
+            final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+            final int nbShards =  randomIntBetween(1, 5);
+            applyClusterState("restore-indices-when-master-" + indexName,
+                clusterState -> addUnassignedShards(clusterState, indexName, nbShards));
+        }
+
+        applyClusterState("demote-current-master", this::demoteMasterNode);
+
+        for (int i = 0; i < randomIntBetween(1, 10); i++) {
+            final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+            final int nbShards =  randomIntBetween(1, 5);
+            applyClusterState("restore-indices-when-no-longer-master-" + indexName,
+                clusterState -> addUnassignedShards(clusterState, indexName, nbShards));
+        }
+
+        assertBusy(() -> {
+            assertThat(snapshotsInfoService.numberOfKnownSnapshotShardSizes(), equalTo(0));
+            assertThat(snapshotsInfoService.numberOfUnknownSnapshotShardSizes(), equalTo(0));
+        });
+    }
+
+    private void applyClusterState(final String reason, final Function<ClusterState, ClusterState> applier) throws Exception {
+        final CompletableFuture<String> future = new CompletableFuture<>();
+        clusterService.getClusterApplierService().onNewClusterState(reason,
```

Comment:
NIT: Could do this in one go with the utilities we have as (also avoid mixing our own concurrency primitives with the JDK's):

```java
        PlainActionFuture.get(future -> clusterService.getClusterApplierService().onNewClusterState(reason,
                () -> applier.apply(clusterService.state()),
                new ClusterApplier.ClusterApplyListener() {
                    @Override
                    public void onSuccess(String source) {
                        future.onResponse(source);
                    }

                    @Override
                    public void onFailure(String source, Exception e) {
                        future.onFailure(e);
                    }
                }));
```

Reply:
Indeed, thanks: 6dcff94a1e851651bf8c954f07276426cbedb970